File: components/brave_wallet/browser/polkadot/polkadot_assets.cc
/* Copyright (c) 2026 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at https://mozilla.org/MPL/2.0/. */

#include "brave/components/brave_wallet/browser/polkadot/polkadot_assets.h"

#include <algorithm>
#include <iterator>
#include <string>
#include <utility>

#include "base/containers/to_vector.h"

namespace brave_wallet {

namespace {

std::string JoinStorageKeys(base::span<const std::string> storage_keys) {
  std::string joined;
  for (const auto& storage_key : storage_keys) {
    if (!joined.empty()) {
      joined.push_back('\n');
    }
    joined.append(storage_key);
  }
  return joined;
}

std::vector<uint8_t> FromRust(const rust::Vec<uint8_t>& vec) {
  return base::ToVector(vec);
}

CxxAssetMetadata ToCxxAssetMetadata(std::array<uint8_t, 16> deposit,
                                    base::span<const uint8_t> name,
                                    base::span<const uint8_t> symbol,
                                    uint8_t decimals,
                                    bool is_frozen) {
  CxxAssetMetadata metadata;
  metadata.deposit = deposit;
  metadata.name = rust::Vec<uint8_t>();
  metadata.name.reserve(name.size());
  std::ranges::copy(name, std::back_inserter(metadata.name));
  metadata.symbol = rust::Vec<uint8_t>();
  metadata.symbol.reserve(symbol.size());
  std::ranges::copy(symbol, std::back_inserter(metadata.symbol));
  metadata.decimals = decimals;
  metadata.is_frozen = is_frozen;
  return metadata;
}

std::array<uint8_t, 16> ToArray(base::span<const uint8_t, 16> deposit) {
  std::array<uint8_t, 16> result = {};
  std::ranges::copy(deposit, result.begin());
  return result;
}

}  // namespace

AssetsList::AssetsList(const AssetsList&) = default;
AssetsList::AssetsList(AssetsList&&) noexcept = default;
AssetsList::~AssetsList() = default;
AssetsList& AssetsList::operator=(const AssetsList&) = default;
AssetsList& AssetsList::operator=(AssetsList&&) noexcept = default;

std::optional<AssetsList> AssetsList::FromStorageKeys(
    base::span<const std::string> storage_keys) {
  auto result =
      parse_assets_list_from_storage_keys(JoinStorageKeys(storage_keys));
  if (!result->is_ok()) {
    return std::nullopt;
  }

  return AssetsList(base::ToVector(result->unwrap()->identifiers));
}

AssetsList::AssetsList(std::vector<uint32_t> identifiers)
    : identifiers_(std::move(identifiers)) {}

AssetMetadata::AssetMetadata(const AssetMetadata& other)
    : AssetMetadata(ToCxxAssetMetadata(ToArray(other.deposit()),
                                       other.name_,
                                       other.symbol_,
                                       other.decimals(),
                                       other.is_frozen()),
                    other.name_,
                    other.symbol_) {}

AssetMetadata::AssetMetadata(AssetMetadata&&) noexcept = default;
AssetMetadata::~AssetMetadata() = default;
AssetMetadata& AssetMetadata::operator=(const AssetMetadata& other) {
  if (this == &other) {
    return *this;
  }

  *this = AssetMetadata(other);
  return *this;
}

AssetMetadata& AssetMetadata::operator=(AssetMetadata&&) noexcept = default;

bool AssetMetadata::operator==(const AssetMetadata& other) const {
  return std::ranges::equal(deposit(), other.deposit()) &&
         name_ == other.name_ && symbol_ == other.symbol_ &&
         decimals() == other.decimals() && is_frozen() == other.is_frozen();
}

std::optional<AssetMetadata> AssetMetadata::FromBytes(
    base::span<const uint8_t> metadata_bytes) {
  auto result = parse_asset_metadata_from_scale(
      ::rust::Slice<const uint8_t>(metadata_bytes));
  if (!result->is_ok()) {
    return std::nullopt;
  }

  auto parsed = result->unwrap();
  return FromFields(parsed->deposit, FromRust(parsed->name),
                    FromRust(parsed->symbol), parsed->decimals,
                    parsed->is_frozen);
}

AssetMetadata AssetMetadata::FromFields(std::array<uint8_t, 16> deposit,
                                        std::vector<uint8_t> name,
                                        std::vector<uint8_t> symbol,
                                        uint8_t decimals,
                                        bool is_frozen) {
  auto metadata =
      ToCxxAssetMetadata(deposit, name, symbol, decimals, is_frozen);
  return AssetMetadata(std::move(metadata), std::move(name), std::move(symbol));
}

const CxxAssetMetadata& AssetMetadata::operator*() const {
  return asset_metadata_;
}

base::span<const uint8_t, 16> AssetMetadata::deposit() const {
  return asset_metadata_.deposit;
}

AssetMetadata::AssetMetadata(CxxAssetMetadata asset_metadata,
                             std::vector<uint8_t> name,
                             std::vector<uint8_t> symbol)
    : asset_metadata_(std::move(asset_metadata)),
      name_(std::move(name)),
      symbol_(std::move(symbol)) {}

}  // namespace brave_wallet

File: components/brave_wallet/browser/internal/polkadot_assets.rs
// Copyright (c) 2026 The Brave Authors. All rights reserved.
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this file,
// You can obtain one at https://mozilla.org/MPL/2.0/.

use crate::polkadot_scale_utils::decode_scale;
use crate::Error;

const ASSETS_STORAGE_MAP_PREFIX_SIZE: usize = 32;
const BLAKE_TWO_128_CONCAT_U32_STORAGE_KEY_SIZE: usize = 20;
const ASSETS_STORAGE_KEY_SIZE: usize =
    ASSETS_STORAGE_MAP_PREFIX_SIZE + BLAKE_TWO_128_CONCAT_U32_STORAGE_KEY_SIZE;
const ASSETS_ASSET_STORAGE_PREFIX: [u8; ASSETS_STORAGE_MAP_PREFIX_SIZE] = [
    0x68, 0x2a, 0x59, 0xd5, 0x1a, 0xb9, 0xe4, 0x8a, 0x8c, 0x8c, 0xc4, 0x18,
    0xff, 0x97, 0x08, 0xd2, 0xd3, 0x43, 0x71, 0xa1, 0x93, 0xa7, 0x51, 0xee,
    0xa5, 0x88, 0x3e, 0x95, 0x53, 0x45, 0x7b, 0x2e,
];

#[cxx::bridge(namespace = brave_wallet)]
mod ffi {
    #[derive(Clone)]
    struct CxxAssetsList {
        identifiers: Vec<u32>,
    }

    #[derive(Clone)]
    struct CxxAssetMetadata {
        deposit: [u8; 16],
        name: Vec<u8>,
        symbol: Vec<u8>,
        decimals: u8,
        is_frozen: bool,
    }

    extern "Rust" {
        type CxxAssetsListResult;

        fn parse_assets_list_from_storage_keys(
            storage_keys: &str,
        ) -> Box<CxxAssetsListResult>;
        fn is_ok(self: &CxxAssetsListResult) -> bool;
        fn error_message(self: &CxxAssetsListResult) -> String;
        fn unwrap(self: &mut CxxAssetsListResult) -> Box<CxxAssetsList>;

        type CxxAssetMetadataResult;

        fn parse_asset_metadata_from_scale(
            metadata_bytes: &[u8],
        ) -> Box<CxxAssetMetadataResult>;
        fn is_ok(self: &CxxAssetMetadataResult) -> bool;
        fn error_message(self: &CxxAssetMetadataResult) -> String;
        fn unwrap(self: &mut CxxAssetMetadataResult) -> Box<CxxAssetMetadata>;

    }
}

use self::ffi::{CxxAssetMetadata, CxxAssetsList};

crate::impl_result!(CxxAssetsList, CxxAssetsListResult);
crate::impl_result!(CxxAssetMetadata, CxxAssetMetadataResult);

fn hex_value(byte: u8) -> Option<u8> {
    match byte {
        b'0'..=b'9' => Some(byte - b'0'),
        b'a'..=b'f' => Some(byte - b'a' + 10),
        b'A'..=b'F' => Some(byte - b'A' + 10),
        _ => None,
    }
}

fn decode_prefixed_hex(input: &str) -> Result<Vec<u8>, Error> {
    let hex = input.strip_prefix("0x").ok_or(Error::InvalidMetadata)?;
    if hex.len() % 2 != 0 {
        return Err(Error::InvalidLength);
    }

    hex.as_bytes()
        .chunks_exact(2)
        .map(|chunk| {
            let high = hex_value(chunk[0]).ok_or(Error::InvalidMetadata)?;
            let low = hex_value(chunk[1]).ok_or(Error::InvalidMetadata)?;
            Ok((high << 4) | low)
        })
        .collect()
}

fn parse_asset_id_from_storage_key(storage_key: &str) -> Result<u32, Error> {
    let bytes = decode_prefixed_hex(storage_key)?;
    if bytes.len() != ASSETS_STORAGE_KEY_SIZE {
        return Err(Error::InvalidLength);
    }
    if bytes[..ASSETS_STORAGE_MAP_PREFIX_SIZE] != ASSETS_ASSET_STORAGE_PREFIX {
        return Err(Error::InvalidMetadata);
    }

    let id_bytes: [u8; 4] = bytes[ASSETS_STORAGE_KEY_SIZE - 4..]
        .try_into()
        .map_err(|_| Error::InvalidLength)?;
    Ok(u32::from_le_bytes(id_bytes))
}

fn parse_assets_list(storage_keys: &str) -> Result<CxxAssetsList, Error> {
    let mut identifiers = Vec::new();
    for storage_key in storage_keys.lines().filter(|key| !key.is_empty()) {
        identifiers.push(parse_asset_id_from_storage_key(storage_key)?);
    }
    Ok(CxxAssetsList { identifiers })
}

fn parse_assets_list_from_storage_keys(storage_keys: &str) -> Box<CxxAssetsListResult> {
    Box::new(CxxAssetsListResult(parse_assets_list(storage_keys)))
}

fn parse_asset_metadata(metadata_bytes: &[u8]) -> Result<CxxAssetMetadata, Error> {
    let mut input = metadata_bytes;
    let deposit: u128 = decode_scale(&mut input)?;
    let name: Vec<u8> = decode_scale(&mut input)?;
    let symbol: Vec<u8> = decode_scale(&mut input)?;
    let decimals: u8 = decode_scale(&mut input)?;
    let is_frozen: bool = decode_scale(&mut input)?;
    if !input.is_empty() {
        return Err(Error::InvalidLength);
    }

    Ok(CxxAssetMetadata {
        deposit: deposit.to_le_bytes(),
        name,
        symbol,
        decimals,
        is_frozen,
    })
}

fn parse_asset_metadata_from_scale(metadata_bytes: &[u8]) -> Box<CxxAssetMetadataResult> {
    Box::new(CxxAssetMetadataResult(parse_asset_metadata(metadata_bytes)))
}
